src/highdicom/legacy/sop.py

```diff
@@ -25,384 +31,506 @@
+        from pydicom.valuerep import DSfloat
+        float_tolerance = 1.0e-5
+
+        def is_equal_float(x1: float, x2: float) -> bool:
+            return abs(x1 - x2) < float_tolerance
+        if type(v1) != type(v2):
+            return False
+        if isinstance(v1, DataElementSequence):
+            for item1, item2 in zip(v1, v2):
+                cls.isequal_dicom_dataset(item1, item2)
+        if not isinstance(v1, MultiValue):
+            v11 = [v1]
+            v22 = [v2]
         else:
-            frame_type_item.FrameVolumeBasedCalculationTechnique = 'MIXED'
-
-        if sop_class_uid == '1.2.840.10008.5.1.4.1.1.4.4':
-            # MR Image Frame Type (M)
-            perframe_item.MRImageFrameTypeSequence = [
-                frame_type_item,
-            ]
-
-        elif sop_class_uid == '1.2.840.10008.5.1.4.1.1.2.2':
-            # CT Image Frame Type (M)
-            perframe_item.CTImageFrameTypeSequence = [
-                frame_type_item,
-            ]
-
-            # CT Pixel Value Transformation (M)
-            pixel_val_transform_item = Dataset()
-            pixel_val_transform_item.RescaleIntercept = ds.RescaleIntercept
-            pixel_val_transform_item.RescaleSlope = ds.RescaleSlope
-            try:
-                pixel_val_transform_item.RescaleType = ds.RescaleType
-            except AttributeError:
-                pixel_val_transform_item.RescaleType = 'US'
-            perframe_item.PixelValueTransformationSequence = [
-                pixel_val_transform_item,
-            ]
-
-        elif sop_class_uid == '1.2.840.10008.5.1.4.1.1.128.1':
-            # PET Image Frame Type (M)
-            perframe_item.PETImageFrameTypeSequence = [
-                frame_type_item,
-            ]
-
-        # Frame VOI LUT (U)
-        try:
-            frame_voi_lut_item = Dataset()
-            frame_voi_lut_item.WindowCenter = ds.WindowCenter
-            frame_voi_lut_item.WindowWidth = ds.WindowWidth
-            perframe_item.FrameVOILUTSequence = [
-                frame_voi_lut_item,
-            ]
-        except AttributeError:
-            pass
-
-        # Referenced Image (C)
-        try:
-            perframe_item.ReferencedImageSequence = \
-                ds.ReferencedImageSequence
-        except AttributeError:
-            pass
+            v11 = v1
+            v22 = v2
+        if len(v11) != len(v22):
+            return False
+        for xx, yy in zip(v11, v22):
+            if isinstance(xx, DSfloat) or isinstance(xx, float):
+                if not is_equal_float(xx, yy):
+                    return False
+            else:
+                if xx != yy:
+                    return False
+        return True
 
-        # Derivation Image (C)
-        try:
-            perframe_item.SourceImageSequence = ds.SourceImageSequence
-        except AttributeError:
-            pass
+    @classmethod
+    def isequal_dicom_dataset(cls, ds1: Dataset, ds2: Dataset) -> bool:
```

Comment:
I wonder to which extent this class and the individual methods are necessary. The *pydicom* library seems to support (at least a subset of) them, for example [pydicom.dataset.Dataset.__eq__](https://github.com/pydicom/pydicom/blob/master/pydicom/dataset.py#L694).

Reply:
You are right. The main reason though for me is to have control over floating-point comparison considering a tolerance.